crates/oxc_transformer/src/common/arrow_function_converter.rs

```diff
@@ -366,6 +380,19 @@ impl<'a> ArrowFunctionConverter<'a> {
                 Ancestor::ArrowFunctionExpressionBody(func) => {
                     return Some(func.scope_id().get().unwrap())
                 }
+                Ancestor::FunctionBody(func) => {
+                    if self.is_async_only() && *func.r#async() &&
+                    // Check the function is part of a method definition or object property,
+                    // if it is, then we treat it as a arrow function.
+                    matches!(
+                        ctx.ancestor(index + 1),
+                        // `class A { async foo() {} }` | `({ async foo() {} })`
+                        Ancestor::MethodDefinitionValue(_) | Ancestor::ObjectPropertyValue(_)
+                    ) {
+                        return Some(func.scope_id().get().unwrap())
+                    }
```

Comment:
We need to check the parent ancestor of the current ancestor here. I didn't thought a better way to do it

Reply:
How about this?

```rs
let mut ancestors = ctx.ancestors();
while let Some(ancestor) = ancestors.next() {
    match ancestor {
        // ...
        Ancestor::FunctionBody(func) => {
            if self.is_async_only() && *func.r#async()
            && matches!(
                ancestors.next().unwrap(),
                Ancestor::MethodDefinitionValue(_) | Ancestor::ObjectPropertyValue(_)
            ) {
                return Some(func.scope_id().get().unwrap());
            }
            return None;
        }
    }
}
unreachable!();
```

Reply:
Looks like better than mine!